azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/AADB2CURL.java

```diff
@@ -0,0 +1,103 @@
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import org.springframework.util.Assert;
+
+import javax.servlet.http.HttpServletRequest;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+
+import static java.lang.String.format;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class AADB2CURL {
+
+    private static final String API_PATTERN = "https://%s.b2clogin.com/%s.onmicrosoft.com/oauth2/v2.0/%s?";
+
+    private static final String PARAMETER_PATTERN = "%s=%s";
+
+    private static final String CLIENT_ID = "client_id";
+
+    private static final String REDIRECT_URL = "redirect_uri";
+
+    private static final String RESPONSE_MODE = "response_mode";
+
+    private static final String RESPONSE_TYPE = "response_type";
+
+    private static final String SCOPE = "scope";
+
+    private static final String STATE = "state";
+
+    private static final String NONCE = "nonce";
+
+    private static final String POLICY = "p";
+
+    private static final String API_TYPE_AUTHORIZE = "authorize";
+
+    private static final String RESPONSE_MODE_QUERY = "query";
+
+    private static final String RESPONSE_TYPE_CODE = "code";
+
+    private static final String RESPONSE_TYPE_ID_TOKEN = "id_token";
+
+    private static final String SCOPE_OPENID = "openid";
+
+    private static String getUUID() {
+        return UUID.randomUUID().toString();
+    }
+
+    private static String getEncodedURL(@NonNull String url) {
+        try {
+            return URLEncoder.encode(url, "utf-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new AADB2CConfigurationException("failed to encode url: " + url, e);
+        }
+    }
+
+    /**
+     * Take state's format as UUID-RequestURI when redirect to sign-in URL.
+     *
+     * @param requestURL from ${@link HttpServletRequest} that user attempt to access.
+     * @return the encoded state String.
+     */
+    private static String getState(String requestURL) {
+        return String.join("-", getUUID(), getEncodedURL(requestURL));
+    }
+
+    /**
+     * Get the openId sign up or sign in redirect URL based on ${@link AADB2CProperties}, encodes the
+     * requestURL and UUID in ${@link AADB2CURL#SCOPE} field.
+     *
+     * @param properties of ${@link AADB2CProperties}
+     * @param requestURL from ${@link HttpServletRequest} that user attempt to access.
+     * @return
+     */
+    public static String getOpenIdSignUpOrSignInUrl(@NonNull AADB2CProperties properties, String requestURL) {
+        Assert.hasText(requestURL, "requestURL should have text.");
+
+        final String endpoint = format(API_PATTERN, properties.getTenant(), properties.getTenant(), API_TYPE_AUTHORIZE);
+        final AADB2CProperties.Policy policy = properties.getPolicies().getSignUpOrSignIn();
+        final List<String> parameters = Arrays.asList(
```

Comment:
Can we use a big fat well-formatted string pattern here? I don't see why we need so many small strings.

Reply:
as variable naming hint, it is `parameters` for redirect URL, for each line is one parameter like `%s=%s`, or I may add some comments for this.
The big format may looks like
`"%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s`, If we leverage this one,  we need to take about 16 parameters to the String.format, I don't think that is a good idea.

Reply:
Leave the choice to you  since this is just personal preference. 😄

Reply:
Thanks, update to big pattern for consideration of more straight forward, as well as readability.